src/websocket_tracker_connection.cpp

```diff
@@ -0,0 +1,490 @@
+/*
+
+Copyright (c) 2019 Paul-Louis Ageneau
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/config.hpp" // for TORRENT_USE_RTC
+
+#if TORRENT_USE_RTC
+
+#include "libtorrent/aux_/escape_string.hpp"
+#include "libtorrent/aux_/session_settings.hpp"
+#include "libtorrent/utf8.hpp"
+#include "libtorrent/io.hpp"
+#include "libtorrent/ip_filter.hpp"
+#include "libtorrent/socket.hpp"
+#include "libtorrent/socket_io.hpp"
+#include "libtorrent/span.hpp"
+#include "libtorrent/tracker_manager.hpp"
+#include "libtorrent/websocket_tracker_connection.hpp"
+
+#include "libtorrent/aux_/disable_warnings_push.hpp"
+#include "boost/json.hpp"
+#include "libtorrent/aux_/disable_warnings_pop.hpp"
+
+#include <boost/system/system_error.hpp>
+
+#include <algorithm>
+#include <cctype>
+#include <cstdio> // for snprintf
+#include <exception>
+#include <functional>
+#include <list>
+#include <locale>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace libtorrent {
+
+using namespace std::placeholders;
+namespace errc = boost::system::errc;
+namespace error = boost::asio::error;
+namespace json = boost::json;
+
+// Convert ISO-8859-1 (aka latin1) input to UTF-8
+std::string from_latin1(span<char const> s) {
```

Comment:
this function should have a unit test. and you may want to `u32.reserve(s.size());`